diff --git a/Makefile b/Makefile
index 038a834..43fe6e3 100644
--- a/Makefile
+++ b/Makefile
@@ -23,6 +23,9 @@ pgbouncer_SOURCES = \
 	src/system.c \
 	src/takeover.c \
 	src/util.c \
+    src/pycall.c \
+    src/route_connection.c \
+    src/rewrite_query.c \
 	src/varcache.c \
 	src/common/base64.c \
 	src/common/saslprep.c \
@@ -49,6 +52,9 @@ pgbouncer_SOURCES = \
 	include/system.h \
 	include/takeover.h \
 	include/util.h \
+    include/pycall.h \
+    include/route_connection.h \
+    include/rewrite_query.h \
 	include/varcache.h \
 	include/common/base64.h \
 	include/common/pg_wchar.h \
@@ -58,19 +64,20 @@ pgbouncer_SOURCES = \
 	include/common/unicode_norm.h \
 	include/common/unicode_norm_table.h
 
-pgbouncer_CPPFLAGS = -Iinclude $(CARES_CFLAGS) $(LIBEVENT_CFLAGS) $(TLS_CPPFLAGS)
+python_CPPFLAGS = -I/usr/include/python2.6 -I/usr/include/python2.6
+pgbouncer_CPPFLAGS = -Iinclude $(CARES_CFLAGS) $(TLS_CPPFLAGS) $(python_CPPFLAGS)
 
 # include libusual sources directly
 AM_FEATURES = libusual
 pgbouncer_EMBED_LIBUSUAL = 1
 
 # docs to install as-is
-dist_doc_DATA = README.md NEWS.md etc/pgbouncer.ini etc/userlist.txt
+dist_doc_DATA = README.md etc/pgbouncer.ini etc/userlist.txt
 
 DISTCLEANFILES = config.mak config.status lib/usual/config.h config.log
 
 DIST_SUBDIRS = doc test
-dist_man_MANS = doc/pgbouncer.1 doc/pgbouncer.5
+dist_man_MANS = 
 
 # files in tgz
 EXTRA_DIST = AUTHORS COPYRIGHT Makefile config.mak.in config.sub config.guess \
@@ -94,8 +101,9 @@ LIBUSUAL_DIST = $(filter-out %/config.h, $(sort $(wildcard \
 		lib/README lib/COPYRIGHT \
 		lib/find_modules.sh )))
 
+python_LDFLAGS = -lpthread -ldl -lutil -lm -lpython2.6 -Xlinker -export-dynamic
 pgbouncer_LDFLAGS := $(TLS_LDFLAGS)
-pgbouncer_LDADD := $(CARES_LIBS) $(LIBEVENT_LIBS) $(TLS_LIBS) $(LIBS)
+pgbouncer_LDADD := $(CARES_LIBS) $(LIBEVENT_LIBS) $(TLS_LIBS) $(LIBS) $(python_LDFLAGS)
 LIBS :=
 
 #
