diff --git a/include/janitor.h b/include/janitor.h
index c4e5ab4..016f93b 100644
--- a/include/janitor.h
+++ b/include/janitor.h
@@ -20,6 +20,7 @@ void janitor_setup(void);
 void config_postprocess(void);
 void resume_all(void);
 void per_loop_maint(void);
+void run_once_to_init(void);
 bool suspend_socket(PgSocket *sk, bool force)  _MUSTCHECK;
 void kill_pool(PgPool *pool);
 void kill_peer_pool(PgPool *pool);
