diff --git a/include/loader.h b/include/loader.h
index 1acbefe..ddb044f 100644
--- a/include/loader.h
+++ b/include/loader.h
@@ -25,3 +25,5 @@ bool parse_user(void *base, const char *name, const char *params) _MUSTCHECK;
 /* user file parsing */
 bool load_auth_file(const char *fn) /* _MUSTCHECK */;
 bool loader_users_check(void) /* _MUSTCHECK */;
+
+PgPool *new_pool_from_db(PgDatabase *db, char *dbname, char *hostname);
