diff --git a/include/objects.h b/include/objects.h
index 0b498dd..7fe471d 100644
--- a/include/objects.h
+++ b/include/objects.h
@@ -37,6 +37,9 @@ extern struct Slab *outstanding_request_cache;
 extern struct Slab *var_list_cache;
 extern struct Slab *server_prepared_statement_cache;
 extern PgPreparedStatement *prepared_statements;
+extern bool fast_switchover;
+/* Store the database info from PgBouncer config to support fast switchover. */
+extern PgDatabase *fast_switchover_db;

 extern unsigned long long int last_pgsocket_id;
 
