diff --git a/include/util.h b/include/util.h
index 387d706..4f7a132 100644
--- a/include/util.h
+++ b/include/util.h
@@ -20,6 +20,16 @@
 #include <usual/string.h>
 #include <usual/cfparser.h>
 
+/*
+ * sets the global writer to NULL for the pool
+ */
+void clear_global_writer(PgPool *pool);
+
+/*
+ * get the global writer, if any from the pool
+ */
+PgPool *get_global_writer(PgPool *pool);
+
 /*
  * logging about specific socket
  */
