diff --git a/include/util.h b/include/util.h
index 35a284b..32e9782 100644
--- a/include/util.h
+++ b/include/util.h
@@ -19,6 +19,16 @@
 #include <usual/logging.h>
 #include <usual/string.h>
 
+/*
+ * sets the global writer to NULL for the pool
+ */
+void clear_global_writer(PgPool *pool);
+
+/*
+ * get the global writer, if any from the pool
+ */
+PgPool *get_global_writer(PgPool *pool);
+
 /*
  * logging about specific socket
  */
