diff --git a/src/client.c b/src/client.c
index fcf7b3b..9cf3b86 100644
--- a/src/client.c
+++ b/src/client.c
@@ -1291,7 +1291,7 @@ static bool handle_client_startup(PgSocket *client, PktHdr *pkt)
 			return false;
 		}
 
-		if (client->pool && !sending_auth_query(client)) {
+		if (client->pool && !sending_auth_query(client) && !get_global_writer(client->pool)) {
 			disconnect_client(client, true, "client re-sent startup pkt");
 			return false;
 		}
@@ -1428,6 +1428,7 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
	int track_outstanding = false;
	PreparedStatementAction ps_action = PS_IGNORE;
	PgClosePacket close_packet;
+	bool in_transaction = false;
 
 	switch (pkt->type) {
	/* one-packet queries */
@@ -1591,8 +1592,11 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
 		client->query_start = get_cached_time();
 	}
 
+	/* xact_start is non-zero if we are inside a transaction */
+	in_transaction = client->xact_start != 0;
+
 	/* remember timestamp of the first query in a transaction */
-	if (!client->xact_start) {
+	if (!in_transaction) {
 		client->pool->stats.xact_count++;
 		client->xact_start = client->query_start;
 	}
@@ -1600,6 +1604,14 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
 	if (client->pool->db->admin)
 		return admin_handle_client(client, pkt);
 
+	/* pgbouncer-rr extensions: query rewrite & client connection routing */
+	if (pkt->type == PqMsg_Query || pkt->type == PqMsg_Parse) {
+		if (!rewrite_query(client, in_transaction, pkt)) {
+			return false;
+		}
+		route_client_connection(client, in_transaction, pkt);
+	}
+
 	/* acquire server */
 	if (!find_server(client))
 		return false;
