diff --git a/src/client.c b/src/client.c
index ac13a48..6057c8a 100644
--- a/src/client.c
+++ b/src/client.c
@@ -1106,7 +1106,7 @@ static bool handle_client_startup(PgSocket *client, PktHdr *pkt)
 			return false;
 		}
 
-		if (client->pool && !client->wait_for_user_conn && !client->wait_for_user) {
+		if (client->pool && !client->wait_for_user_conn && !client->wait_for_user && !get_global_writer(client->pool)) {
 			disconnect_client(client, true, "client re-sent startup pkt");
 			return false;
 		}
@@ -1234,6 +1234,7 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
 	int track_outstanding = false;
 	PreparedStatementAction ps_action = PS_IGNORE;
 	PgClosePacket close_packet;
+	bool in_transaction = false;
 
 	switch (pkt->type) {
 	/* one-packet queries */
@@ -1397,8 +1398,11 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
 		client->query_start = get_cached_time();
 	}
 
+	/* xact_start is non-zero if we are inside a transaction */
+	in_transaction = client->xact_start != 0;
+
 	/* remember timestamp of the first query in a transaction */
-	if (!client->xact_start) {
+	if (!in_transaction) {
 		client->pool->stats.xact_count++;
 		client->xact_start = client->query_start;
 	}
@@ -1406,6 +1410,14 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
 	if (client->pool->db->admin)
 		return admin_handle_client(client, pkt);
 
+	/* pgbouncer-rr extensions: query rewrite & client connection routing */
+	if (pkt->type == 'Q' || pkt->type == 'P') {
+		if (!rewrite_query(client, in_transaction, pkt)) {
+			return false;
+		}
+		route_client_connection(client, in_transaction, pkt);
+	}
+
 	/* acquire server */
 	if (!find_server(client))
 		return false;
