--- ../pgbouncer/src/client.c	2020-04-26 15:20:51.660323326 -0400
+++ ./src/client.c	2020-04-26 15:19:04.085959392 -0400
@@ -480,16 +480,16 @@ static bool decide_startup_pool(PgSocket
 		} else if (strcmp(key, "user") == 0) {
 			slog_debug(client, "got var: %s=%s", key, val);
 			username = val;
 		} else if (strcmp(key, "application_name") == 0) {
 			set_appname(client, val);
 			appname_found = true;
-		} else if (varcache_set(&client->vars, key, val)) {
-			slog_debug(client, "got var: %s=%s", key, val);
 		} else if (strlist_contains(cf_ignore_startup_params, key)) {
 			slog_debug(client, "ignoring startup parameter: %s=%s", key, val);
+		} else if (varcache_set(&client->vars, key, val)) {
+			slog_debug(client, "got var: %s=%s", key, val);
 		} else {
 			slog_warning(client, "unsupported startup parameter: %s=%s", key, val);
 			disconnect_client(client, true, "unsupported startup parameter: %s", key);
 			return false;
 		}
 	}
@@ -861,12 +861,20 @@ static bool handle_client_work(PgSocket
 		client->xact_start = client->query_start;
 	}
 
 	if (client->pool->db->admin)
 		return admin_handle_client(client, pkt);
 
+        /* pgbouncer-rr extensions: query rewrite & client connection routing */
+        if (pkt->type == 'Q' || pkt->type == 'P') {
+                if (!rewrite_query(client, pkt)) {
+                        return false;
+                }
+                route_client_connection(client, pkt);
+        }
+
 	/* acquire server */
 	if (!find_server(client))
 		return false;
 
 	/* postpone rfq change until certain that client will not be paused */
 	if (rfq_delta) {
