--- a/src/client.c
+++ b/src/client.c
@@ -446,10 +446,10 @@ static bool decide_startup_pool(PgSocket
 		} else if (strcmp(key, "application_name") == 0) {
 			set_appname(client, val);
 			appname_found = true;
-		} else if (varcache_set(&client->vars, key, val)) {
-			slog_debug(client, "got var: %s=%s", key, val);
 		} else if (strlist_contains(cf_ignore_startup_params, key)) {
 			slog_debug(client, "ignoring startup parameter: %s=%s", key, val);
+		} else if (varcache_set(&client->vars, key, val)) {
+			slog_debug(client, "got var: %s=%s", key, val);
 		} else {
 			slog_warning(client, "unsupported startup parameter: %s=%s", key, val);
 			disconnect_client(client, true, "unsupported startup parameter: %s", key);
@@ -678,6 +678,14 @@ static bool handle_client_work(PgSocket
 	if (client->pool->db->admin)
 		return admin_handle_client(client, pkt);
 
+        /* pgbouncer-rr extensions: query rewrite & client connection routing */
+        if (pkt->type == 'Q' || pkt->type == 'P') {
+                if (!rewrite_query(client, pkt)) {
+                        return false;
+                }
+                route_client_connection(client, pkt);
+        }
+
 	/* acquire server */
 	if (!find_server(client))
 		return false;
