diff --git a/src/client.c b/src/client.c
index acaa9a1..1418e4e 100644
--- a/src/client.c
+++ b/src/client.c
@@ -859,6 +859,14 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
 
 	if (client->pool->db->admin)
 		return admin_handle_client(client, pkt);
+    
+    /* pgbouncer-rr extensions: query rewrite & client connection routing */
+    if (pkt->type == 'Q' || pkt->type == 'P') {
+        if (!rewrite_query(client, pkt)) {
+            return false;
+        }
+        route_client_connection(client, pkt);
+    }
 
 	/* acquire server */
 	if (!find_server(client))
