--- ../pgbouncer/src/main.c 2020-04-26 15:20:51.662323314 -0400
+++ ./src/main.c    2020-04-26 15:19:21.995853494 -0400
@@ -160,6 +160,11 @@
 int cf_log_pooler_errors;
 int cf_application_name_add_host;

+/* pgbouncer-rr  extension */
+char *cf_routing_rules_py_module_file;
+char *cf_rewrite_query_py_module_file;
+char *cf_rewrite_query_disconnect_on_failure;
+
 int cf_client_tls_sslmode;
 char *cf_client_tls_protocols;
 char *cf_client_tls_ca_file;
@@ -220,6 +225,10 @@
 static const struct CfKey bouncer_params [] = {
 CF_ABS("admin_users", CF_STR, cf_admin_users, 0, ""),
 CF_ABS("application_name_add_host", CF_INT, cf_application_name_add_host, 0, "0"),
+/* pgbouncer-rr extensions */
+CF_ABS("routing_rules_py_module_file", CF_STR, cf_routing_rules_py_module_file, 0, "not_enabled"),
+CF_ABS("rewrite_query_py_module_file", CF_STR, cf_rewrite_query_py_module_file, 0, "not_enabled"),
+CF_ABS("rewrite_query_disconnect_on_failure", CF_STR, cf_rewrite_query_disconnect_on_failure, 0, "false"),
 CF_ABS("auth_file", CF_STR, cf_auth_file, 0, NULL),
 CF_ABS("auth_hba_file", CF_STR, cf_auth_hba_file, 0, ""),
 CF_ABS("auth_query", CF_STR, cf_auth_query, 0, "SELECT usename, passwd FROM pg_shadow WHERE usename=$1"),