diff --git a/src/pktbuf.c b/src/pktbuf.c
index 7dfec56..c4a6080 100644
--- a/src/pktbuf.c
+++ b/src/pktbuf.c
@@ -65,6 +65,37 @@ PktBuf *pktbuf_dynamic(int start_len)
 	return buf;
 }
 
+PktBuf *pktbuf_copy(PktBuf *orig)
+{
+	PktBuf *buf = zmalloc(sizeof(*orig));
+	if (!buf)
+		return NULL;
+
+	buf->ev = zmalloc(sizeof(*orig->ev));
+	if (!buf->ev) {
+		pktbuf_free(buf);
+		return NULL;
+	}
+	buf->buf = zmalloc(orig->buf_len);
+	if (!buf->buf) {
+		pktbuf_free(buf);
+		return NULL;
+	}
+
+	memcpy(buf->buf, orig->buf, orig->buf_len);
+	memcpy(buf->ev, orig->ev, sizeof(*orig->ev));
+
+	buf->buf_len = orig->buf_len;
+	buf->write_pos = orig->write_pos;
+	buf->pktlen_pos = orig->pktlen_pos;
+	buf->send_pos = orig->send_pos;
+	buf->failed = orig->failed;
+	buf->sending = orig->sending;
+	buf->fixed_buf = orig->fixed_buf;
+
+	return buf;
+}
+
 void pktbuf_reset(struct PktBuf *pkt)
 {
 	pkt->failed = false;
